rust/kernel/cpu.rs

```diff
@@ -0,0 +1,127 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! CPU utilities for SMP systems.
+//!
+//! C header: [`include/linux/smp.h`](../../../../include/linux/smp.h).
+//!
+//! Note: This module is a work in progress.
+
+use crate::{bindings, error::code::*, prelude::*};
+use core::marker::PhantomData;
+
+/// Represents an SMP CPU by its zero-based kernel processor id.
+///
+/// Cpu values can be safely passed between threads (and across physical CPUs) as they are
+/// merely an integer index representing a CPU. All actual interaction with CPUs is late-
+/// bound.
+#[derive(Copy, Clone)]
+pub struct Cpu(i32);
+
+struct FnCall<F, R>
+where
+    F: FnOnce() -> R,
+{
+    out: Option<R>,
+    func: Option<F>,
+}
+
+unsafe extern "C" fn call_fn<F, R>(fn_box: *mut core::ffi::c_void)
+where
+    F: FnOnce() -> R,
+{
+    // SAFETY: The casting here is the invesrse of the cast to *mut c_void that occurs in Cpu::call
+    // below. That call waits for this call to complete ensuring fn_call is live for the duration.
+    let fn_call = unsafe { &mut *(fn_box as *mut FnCall<F, R>) };
+    let func = fn_call.func.take().unwrap();
```

Comment:
Maybe this is a premature optimization. But `call_fn` is called inside the irq handler, and usually we want an irq handler to be as small as possible. However, by `Send`ing the closure, we actually place the `drop` of the closure inside an irq handler, which adds more work for the irq handler.

Although I doubt someone would put a "large" object in an IPI closure, but still it's always good to save some cpu cycles in irq handlers.

I understand that `FnOnce()` is used, because we want to only make the IPI closure called once. But I don't that matters a lot here, because the implementation guarantees it's only called once. Maybe we can use `Fn` here?

Reply:
It should be noted that we are calling a `FnOnce()` which consumes itself when called, so there are no calls to any drop glues here inside the function.

Whether any dropping is done when the function completes is another issue -- I would say that the decision for that should be leave to the user who supplies the closure.
* If the closure requires ownership (and thus needs moving and dropping), restricting to `Fn` here will require the user to use the `Option` and `take` trick themselves, which is not good;
* If the closure doesn't require ownership, and `move` is not specified, then the closure doesn't need dropping anyway so there is no advantage for `Fn` anyway;
* If the closure doesn't require ownership but `move` is used (and thus requires `Drop`), then the user can pass in `&closure` or `&mut closure` (`FnOnce` is implemented for `&impl Fn` and `&mut impl FnOnce`) and they will be drop-free; or maybe the user should just drop `move`

Reply:
> It should be noted that we are calling a `FnOnce()` which consumes itself when called, so there are no calls to any drop glues here inside the function.
> 

Agreed.

> Whether any dropping is done when the function completes is another issue -- I would say that the decision for that should be leave to the user who supplies the closure.
> 
> * If the closure requires ownership (and thus needs moving and dropping), restricting to `Fn` here will require the user to use the `Option` and `take` trick themselves, which is not good;
> * If the closure doesn't require ownership, and `move` is not specified, then the closure doesn't need dropping anyway so there is no advantage for `Fn` anyway;
> * If the closure doesn't require ownership but `move` is used (and thus requires `Drop`), then the user can pass in `&closure` or `&mut closure` (`FnOnce` is implemented for `&impl Fn` and `&mut impl FnOnce`) and they will be drop-free; or maybe the user should just drop `move`

Ah, just learn that `FnOnce` is implemented for `&impl Fn` and `&mut impl FnMut`, thanks!

Then I'm less worried ;-) 

In real world, I don't think users will write IPI closures that require ownership, but it's better that we add some comments for `Cpu::call` to warn users about the overhead. For example:

```rust
/// Be careful about giving any ownership to the IPI closure because it will reach the end of life in
/// the irq handler and `drop` anything it owns. If anything is `move`d in an IPI closure, please
/// consider passing a `& impl Fn` or `&mut FnMut` as `func` to avoid dropping inside irq handlers.
```

Thoughts?

Reply:
I incorporated a slightly wordier version of this comment that I think captures the same sentiment, but also highlights that the caller may want to avoid dropping in the interrupt handler as doing so may be expensive, time-wise.